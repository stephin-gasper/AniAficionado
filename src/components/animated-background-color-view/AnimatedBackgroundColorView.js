import Animated, {
  useSharedValue,
  useAnimatedStyle,
  withTiming,
  Easing,
} from 'react-native-reanimated';
import React, {useEffect} from 'react';
import PropTypes from 'prop-types';

const AnimatedBackgroundColorView = ({
  children,
  style,
  initialColor,
  activeColor,
}) => {
  const backgroundColor = useSharedValue(initialColor);

  const animatedStyle = useAnimatedStyle(
    () => ({
      backgroundColor: withTiming(backgroundColor.value, {
        duration: 1000,
        easing: Easing.out(Easing.exp),
      }),
    }),
    [activeColor],
  );

  useEffect(() => {
    if (backgroundColor.value !== activeColor) {
      backgroundColor.value = activeColor;
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [activeColor]);

  return (
    <Animated.View
      testID="animatedBackgroundColorView"
      style={[style, animatedStyle]}>
      {children}
    </Animated.View>
  );
};

AnimatedBackgroundColorView.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.node,
    PropTypes.func,
    PropTypes.elementType,
  ]).isRequired,
  style: PropTypes.oneOfType([PropTypes.array, PropTypes.object]),
  initialColor: PropTypes.string.isRequired,
  activeColor: PropTypes.string.isRequired,
};

AnimatedBackgroundColorView.defaultProps = {
  style: {},
};

export default AnimatedBackgroundColorView;
