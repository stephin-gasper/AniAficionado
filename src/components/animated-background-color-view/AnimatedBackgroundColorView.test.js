import React from 'react';
import {View as MockView} from 'react-native';
import {render} from 'test/test-utils';
import {
  withReanimatedTimer,
  advanceAnimationByTime,
  getAnimatedStyle,
} from 'react-native-reanimated/lib/reanimated2/jestUtils';

import AnimatedBackgroundColorView from './AnimatedBackgroundColorView';

describe('<AnimatedBackgroundColorView />', () => {
  const defaultProps = {
    initialColor: '#ffffff',
    activeColor: '#ffffff',
  };
  const RenderedComponent = props => (
    <AnimatedBackgroundColorView {...defaultProps} {...props}>
      <MockView testID="childComponent" />
    </AnimatedBackgroundColorView>
  );

  it('should render correctly', () => {
    withReanimatedTimer(() => {
      expect(render(<RenderedComponent />)).toMatchSnapshot();
    });
  });

  it('should additional style when style prop is present', () => {
    withReanimatedTimer(async () => {
      expect(
        render(<RenderedComponent style={{width: 100}} />),
      ).toMatchSnapshot();
    });
  });

  it('should change background color when activeColor changes', () => {
    withReanimatedTimer(() => {
      const {getByTestId, update} = render(<RenderedComponent />);

      const animatedView = getByTestId('animatedBackgroundColorView');
      expect(getAnimatedStyle(animatedView).backgroundColor).toBe('#ffffff');

      update(<RenderedComponent activeColor="#000000" />);

      // middle animation state
      advanceAnimationByTime(500);
      expect(getAnimatedStyle(animatedView).backgroundColor).toBe(
        'rgba(7, 7, 7, 1)',
      );

      // end animation state
      advanceAnimationByTime(600);
      expect(getAnimatedStyle(animatedView).backgroundColor).toBe(
        'rgba(0, 0, 0, 1)',
      );
    });
  });
});
