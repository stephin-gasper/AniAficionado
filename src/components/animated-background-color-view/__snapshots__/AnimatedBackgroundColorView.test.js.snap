// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AnimatedBackgroundColorView /> should additional style when style prop is present 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "backgroundColor": "#ffffff",
      },
    }
  }
  collapsable={false}
  style={
    {
      "backgroundColor": "#ffffff",
      "width": 100,
    }
  }
  testID="animatedBackgroundColorView"
>
  <View
    testID="childComponent"
  />
</View>
`;

exports[`<AnimatedBackgroundColorView /> should render correctly 1`] = `
<View
  animatedStyle={
    {
      "value": {
        "backgroundColor": "#ffffff",
      },
    }
  }
  collapsable={false}
  style={
    {
      "backgroundColor": "#ffffff",
    }
  }
  testID="animatedBackgroundColorView"
>
  <View
    testID="childComponent"
  />
</View>
`;
